### PR TITLE
[REVIEW] Upgrade `pandas` version in `cudf`

### PR DESCRIPTION
**Pull request by galipremsagar**
Fixes: #9125 

This PR:

- [x] Upgrades pandas to its latest version - `1.3.2`
- [x] Fixes multiple hidden issues in `cudf` that were fixed as part of the latest `pandas` release.
- [x] Introduces `xfail` at multiple places which are essentially regressions in the latest `pandas` release.
- [x] Split up and refactor some `pytests` that will help us determine failures easily and are maintainable.

### CODE REVIEW

**Comment by galipremsagar** [comment]
rerun tests

File: ci/gpu/build.sh
Comment (galipremsagar) [inline ci/gpu/build.sh:87]:
```suggestion
gpuci_mamba_retry install -y pandas=1.3.2 arrow-cpp=5.0.0 pyarrow=5.0.0=*cuda
```

File: ci/gpu/build.sh
Comment (galipremsagar) [inline ci/gpu/build.sh:86]:
```suggestion
gpuci_mamba_retry remove --force rapids-build-env rapids-notebook-env
```

File: ci/gpu/build.sh
Comment (galipremsagar) [inline ci/gpu/build.sh:87]:
```suggestion
gpuci_mamba_retry install -y pandas=1.3.2 arrow-cpp=5.0.0 pyarrow=5.0.0=*cuda gmock cython=0.29.24
```

File: ci/gpu/build.sh
Comment (galipremsagar) [inline ci/gpu/build.sh:87]:
```suggestion
gpuci_mamba_retry install -y pandas=1.3.2 arrow-cpp=5.0.0 pyarrow=5.0.0=*cuda gmock cython=0.29.24 pytest pytest-benchmark pytest-xdist
```

File: ci/gpu/build.sh
Comment (galipremsagar) [inline ci/gpu/build.sh:87]:
```suggestion
gpuci_mamba_retry install -y pandas=1.3.2 arrow-cpp=5.0.0 pyarrow=5.0.0=*cuda gmock cython=0.29.24 pytest pytest-benchmark pytest-xdist pytest-cov pytest-timeout
```

File: ci/gpu/build.sh
Comment (galipremsagar) [inline ci/gpu/build.sh:87]: A temporary change which will be reverted once @rapidsai/ops-codeowners merges https://github.com/rapidsai/integration/pull/345

Comment (galipremsagar) [reply ci/gpu/build.sh]: Reverted.

**Review by Ethyling — APPROVED** [review]

File: python/cudf/cudf/tests/test_dataframe.py
Comment (shwina) [inline python/cudf/cudf/tests/test_dataframe.py:1245]: I think in situations like this, it's cleaner to use `assert_eq(expect, got, check_dtypes=False)` -- or does that not achieve the same thing?

File: python/cudf/cudf/tests/test_dataframe.py
Comment (shwina) [inline python/cudf/cudf/tests/test_dataframe.py:1767]: Not really understanding the changes in this test -- can we add a comment to explain what's going on?

File: python/cudf/cudf/tests/test_index.py
Comment (shwina) [inline python/cudf/cudf/tests/test_index.py:1869]: We could just use `check_dtypes=False` here too.

**Review by shwina — COMMENTED** [review]
This is invaluable work, @galipremsagar!

File: python/cudf/cudf/tests/test_dataframe.py
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_dataframe.py:1767]: Sure, so this was a change in behavior incase of `bool` column:

Until pandas `1.2.5`, setting a `nan` to bool column would result in conversion to `float`:
```python
s = pd.Series([True, False, True])
s
0     True
1    False
2     True
dtype: bool
s[[2]] = np.nan
s
0    1.0
1    0.0
2    NaN
dtype: float64
```

Now that's fixed to conversion to an `object` dtype:

```python
>>> s = pd.Series([True, False, True])
>>> s
0     True
1    False
2     True
dtype: bool
>>> s[[2]] = np.nan
>>> s
0     True
1    False
2      NaN
dtype: object
```

So in `cudf` we change `object` dtype to `str` type - for which there is no `transpose` implemented yet.

Hence the fix I have made in this pytest is to correctly cast the data to pandas nullable types and perform a transpose

Comment (shwina) [reply python/cudf/cudf/tests/test_dataframe.py:1767]: Ahh, thanks! Let's add a comment in the code here for our future selves?

File: python/cudf/cudf/tests/test_index.py
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_index.py:1869]: `assert_index_equal` doesn't seem to be having that parameter. The closest thing that skips `type` checking is `exact` param: https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_index_equal.html

File: python/cudf/cudf/tests/test_dataframe.py
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_dataframe.py]: Yup, changed this.

File: python/cudf/cudf/tests/test_dataframe.py
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_dataframe.py:1767]: Done 👍

**Review by shwina — APPROVED** [review]

**Comment by galipremsagar** [comment]
@shwina I made a trivial fix to a failure being caused by latest `dask` upstream changes: https://github.com/rapidsai/cudf/pull/9147/commits/1ded3a3ca99bd59d38b749f043a16e8879b08c9a

This was to match pandas `get_slice_bound` function signature.

**Review by shwina — APPROVED** [review]

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
@gpucibot merge

**Pull request merged**
